stdlib/universe/moving_average.go

```diff
@@ -0,0 +1,444 @@
+	return &movingAverageTransformation{
+		d:       d,
+		cache:   cache,
+		n:       spec.N,
+		columns: spec.Columns,
+	}
+}
+
+func (t *movingAverageTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func (t *movingAverageTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("moving average found duplicate table with key: %v", tbl.Key())
+	}
+	timeIdx := -1
+	cols := tbl.Cols()
+	doMovingAverage := make([]bool, len(cols))
+	for j, c := range cols {
+		found := false
+		for _, label := range t.columns {
+			if c.Label == label {
+				if c.Type != flux.TInt && c.Type != flux.TUInt && c.Type != flux.TFloat {
+					return fmt.Errorf("cannot take moving average of column %s (type %s)", c.Label, c.Type.String())
+				}
+				found = true
+				break
+			}
+		}
+		if c.Label == execute.DefaultTimeColLabel {
+			timeIdx = j
+		}
+
+		if found {
+			mac := c
+			mac.Type = flux.TFloat
+			_, err := builder.AddCol(mac)
+			if err != nil {
+				return err
+			}
+			doMovingAverage[j] = true
+		} else {
+			_, err := builder.AddCol(c)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if timeIdx < 0 {
+		return fmt.Errorf("no column %q exists", execute.DefaultTimeColLabel)
+	}
+
+	return tbl.Do(func(cr flux.ColReader) error {
+		if cr.Len() == 0 {
+			return nil
+		}
+
+		for j, c := range cr.Cols() {
+			var err error
+			switch c.Type {
+			case flux.TBool:
+				err = t.passThroughBool(cr.Times(timeIdx), cr.Bools(j), builder, j)
+			case flux.TInt:
+				err = t.doInt(cr.Times(timeIdx), cr.Ints(j), builder, j, doMovingAverage[j])
+			case flux.TUInt:
+				err = t.doUInt(cr.Times(timeIdx), cr.UInts(j), builder, j, doMovingAverage[j])
+			case flux.TFloat:
+				err = t.doFloat(cr.Times(timeIdx), cr.Floats(j), builder, j, doMovingAverage[j])
+			case flux.TString:
+				err = t.passThroughString(cr.Times(timeIdx), cr.Strings(j), builder, j)
+			case flux.TTime:
+				err = t.passThroughTime(cr.Times(timeIdx), cr.Times(j), builder, j)
+			}
+
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func (t *movingAverageTransformation) UpdateWatermark(id execute.DatasetID, mark execute.Time) error {
+	return t.d.UpdateWatermark(mark)
+}
+
+func (t *movingAverageTransformation) UpdateProcessingTime(id execute.DatasetID, pt execute.Time) error {
+	return t.d.UpdateProcessingTime(pt)
+}
+
+func (t *movingAverageTransformation) Finish(id execute.DatasetID, err error) {
+	t.d.Finish(err)
+}
+
+func (t *movingAverageTransformation) passThroughBool(ts *array.Int64, vs *array.Boolean, b execute.TableBuilder, bj int) error {
+	l := vs.Len()
+	for i := int(t.n - 1); i < l; i++ {
+		if vs.IsValid(i) {
+			if err := b.AppendBool(bj, vs.Value(i)); err != nil {
+				return err
+			}
+		} else {
+			if err := b.AppendNil(bj); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
```

Comment:
This can be done like this:

```go
func (t *movingAverageTransformation) passThroughBool(vs *array.Boolean, b execute.TableBuilder, bj int) error {
	s := arrow.BoolSlice(vs, int(t.n-1), vs.Len())
	defer s.Release()
	return b.AppendBools(bj, s)
}
```

Note that `ts` is useless in the method signature.